text/prod-loader.md

```diff
@@ -0,0 +1,133 @@
+
+The goal of this change will be to:
+
+* Limit the amount of Steal code that needs to run in production to under 10k. I need to compare to other bundlers to see if this is achievable or not, but my guess is that it is.
+* Support circular dependencies.
+* Support progressive loading.
+* Support loading as async script tags so that multiple bundles can be loaded in parallel:
+
+```html
+<script async src="./path/to/prod.js"></script>
+<script async src="./path/to/bundle-a.js"></script>
+<script async src="./path/to/bundle-b.js"></script>
+```
+
+# Impact
+
+This will be small change to the affecting repositories and most work will be done in a new project. This would reflect a *minor* version change as it would introduce a new feature with no backwards compatibility concerns.  In the future this loader could become the default that steal-tools uses, which would require a *major* version change but there are no plans to do that until it has been used in the wild for a bit behind a flag.
+
+The following projects *will* be affected:
+
+* **transpile** - In the below design I lay out a new bundling format that this new loader will use. Transpile will need to be updated in order to transpile to this format.
+* **steal-tools** - Will need to be updated to accept a flag that will use this new loader.
+
+# Design
+
+> Note that API names here aren't important and can change, I'm using steal as the example.
+
+Since the goal of this RFC is a loader that is *as small as possible* it will only support the minimal amount of configuration needed to for progressive loading bundles.
+
+## Module ids replace module names
+
+**transpile** will transpile to a different format, let's call it **_steal_** for now, and will omit something like:
+
+```js
+steal([0, 7, 12, 34], function(exports, bar, baz, qux){
+
+});
+```
+
+Notice that instead of module names there are integers. This is to save space by omitting the long strings. During the build each module name will be given an integer id.
+
+Also notice that the id of **0** is a special `exports` module. This mirrors the `exports` in AMD and allows for compatibility with ES module circular dependencies.
+
+## bundles configuration
+
+Steal has [bundles](http://stealjs.com/docs/System.bundles.html) configuration that looks something like this:
+
+```js
+System.bundles = {
+  "bundles/donejs-chat/home": [
+    "can-map@3.0.0#bubble",
+    "foo-bar@1.0.0#baz"
+  ]
+};
+```
+
+To resolve a moduleName to a bundle Steal needs to loop over each of these keys and perform an `indexOf` to see if the moduleName is part of the bundle.
+
+We can make this more efficient by having the map be module ids to bundle ids. 
+
+```js
+steal.bundles = {
+  3: 144,
+  34: 144,
+  22: 145
+};
+steal.paths = {
+  144: "bundles/donejs-chat/home.js"
+};
+```
+
+## dynamic module name configuration
+
+To support progressive loading we need to be able to map module identifiers to the module id. We can support config like:
+
+```js
```

Comment:
If you're going to have to duplicate these module names elsewhere, it's more likely gzip will mean this doesn't benefit anything.